docs/design/features/readytorun-elf-support-roadmap.md

```diff
@@ -0,0 +1,170 @@
+  paraller 32-bit and 64-bit variants. It might be useful to logically unify them
+  by internally using longs and just have virtual de-serializers that would have
+  something like ReadNativeInt representing 32-bit or 64-bit integer based on
+  the bitness.
+  
+  We probably don't need to support ELF to the fullest, e.g. endianness doesn't
+  matter as our JIT and runtime doesn't support any high-endian CPU's (ARM has
+  technically switchable endianness but we only support the little variant).
+
+* Create a simplified counterpart of System.Reflection.PortableExecutable to
+  support ELF, again somewhere under *src/coreclr/src/tools/Common/Internal/Runtime*.
+  At some point we may consider porting it to libraries but initially it's better
+  to keep that separate.
+  
+  We should also double-check whether mono doesn't already have something in this
+  vein we could reuse or leverage. We basically need to be able to read / write
+  the main ELF header, the program headers, the section headers and the symbols.
+  This is junior level task estimated at 3 man-weeks including creating a validation
+  test that should use the designed API to rewrite an arbitrary ELF executable and
+  verify that existing tools like readelf properly parse the file.
+  
+  It doesn't need to have 100% fidelity but we need to verify it's not distorted
+  and that the main components like targeting OS / architecture, pointers to the
+  sections and symbols are kept intact. The transcript from the dumping tool
+  should be made part of the output from the regression test in order to make
+  potential regressions easy to diagnose.
+  
+  We should also try to figure out whether we might be able to implement the
+  section builder logic in a similar manner for both PE and ELF builder - this
+  would probably require creating a wrapper over the PE builder but it would be
+  still easier than having two completely different code paths for composing the
+  sections. We'll need to address minute differences like alignment considerations
+  as on Linux **file offset - RVA** deltas need to be page-aligned.
+
+* Virtualizing R2RDump header parsing to add ELF support. This should be
+  junior-level task estimated at 1 man-week of work. It should check for the
+  well-known signature magic values used by PE / ELF files to identify the format
+  of the input file (both for the "main" and reference input files) and use
+  proper logic to locate the R2R header, COR header / MetadatReader and the logic
+  to map RVA's to file offsets.
+
+* Virtualizing Crossgen2 SectionBuilder / R2RPEBuilder: this is senior / principal
+  work estimated at 3 man-weeks to refactor section building to be independent on
+  the choice of PE / ELF format and use the newly created ELF API to [re]construct
+  the necessary headers and create a fully functional ELF executable.
+
+  At this point this will probably still be just a dynamic library, not a
```

Comment:
For this, crossgen needs to emit `.o` file and the final binary needs to be produced by platform linker (ie `ld` on Linux).

Reply:
OK, I guess this means we basically need to be able to generate an object or executable flavor of the ELF file. At least I hope they differ less than obj and exe on Windows :-).

Reply:
Generating `.o` from crossgen and linking later would be very useful feature -- this will enable building native linked single-exe apps. 

However, how will cross compilation of apps work? We'll need cross-platform native tool chains (ex: `ld` on Windows)